pkg/firewalld/firewalld.go

```diff
@@ -0,0 +1,108 @@
+package firewalld
+
+import (
+	"fmt"
+
+	"github.com/godbus/dbus"
+)
+
+type Interface struct {
+	obj         *dbus.Object
+	reloadFuncs []func()
+}
+
+const (
+	IPv4 = "ipv4"
+	IPv6 = "ipv6"
+)
+
+const (
+	firewalldName      = "org.fedoraproject.FirewallD1"
+	firewalldPath      = "/org/fedoraproject/FirewallD1"
+	firewalldInterface = "org.fedoraproject.FirewallD1"
+)
+
+func New() *Interface {
+	bus, err := dbus.SystemBus()
+	if err != nil {
+		return &Interface{}
+	}
+
+	fw := &Interface{
+		obj: bus.Object(firewalldName, dbus.ObjectPath(firewalldPath)),
+	}
+
+	go fw.dbusSignalHandler(bus)
+
+	return fw
+}
+
+func (fw *Interface) IsRunning() bool {
+	if fw.obj == nil {
+		return false
+	}
+
+	var zone string
+	err := fw.obj.Call(firewalldInterface+".getDefaultZone", 0).Store(&zone)
+	return err == nil
+}
+
+func (fw *Interface) AddRule(ipv, table, chain string, priority int, rule []string) error {
+	return fw.obj.Call(firewalldInterface+".direct.addRule", 0, ipv, table, chain, int32(priority), rule).Store()
+}
+
+func (fw *Interface) EnsureRule(ipv, table, chain string, priority int, rule []string) error {
+	var exists bool
+
+	err := fw.obj.Call(firewalldInterface+".direct.queryRule", 0, ipv, table, chain, int32(priority), rule).Store(&exists)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return nil
+	}
+
+	return fw.AddRule(ipv, table, chain, priority, rule)
+}
+
+func (fw *Interface) AddReloadFunc(reloadFunc func()) {
+	fw.reloadFuncs = append(fw.reloadFuncs, reloadFunc)
+}
+
+func (fw *Interface) dbusSignalHandler(bus *dbus.Conn) {
```

Comment:
That was my original plan, until I remembered that openshift-sdn doesn't depend on/include kubernetes...

If that changes, then yes, we could use kutil.iptables.AddReloadFunc() instead of this